### PR TITLE
plugins/wrapping: init

### PR DESCRIPTION
**Pull request by ZainKergaye**
Add new plugin: wrapping 

This is my first time really trying to contribute to a tool I daily drive. Please let me know what I am doing wrong. I think the contributing docs are a little confusing for a newcomer. 

Thanks!

### CODE REVIEW

File: plugins/by-name/wrapping/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  inherit (lib.nixvim) defaultNullOpts;
+in
+lib.nixvim.neovim-plugin.mkNeovimPlugin {
+  name = "wrapping";
+  originalName = "wrapping.nvim";
+  package = pkgs.vimPlugins.wrapping-nvim;
+
+  maintainers = [ lib.maintainers.ZainKergaye ];
+
+  # Optionally, explicitly declare some options. You don't have to.
```
Comment (GaetanLepage) [inline plugins/by-name/wrapping/default.nix:16]:
```suggestion
```

File: plugins/by-name/wrapping/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  settingsOptions.notify_on_switch = defaultNullOpts.mkBool true ''
+       By default, wrapping.nvim will output a message to the
+    command line when the hard or soft mode is set.
+  '';
+
+  # Optionally, provide an example for the `settings` option.
```
Comment (GaetanLepage) [inline plugins/by-name/wrapping/default.nix:22]:
```suggestion
```

File: plugins/by-name/wrapping/default.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  originalName = "wrapping.nvim";
+  package = pkgs.vimPlugins.wrapping-nvim;
+
+  maintainers = [ lib.maintainers.ZainKergaye ];
+
+  # Optionally, explicitly declare some options. You don't have to.
+  settingsOptions.notify_on_switch = defaultNullOpts.mkBool true ''
+       By default, wrapping.nvim will output a message to the
+    command line when the hard or soft mode is set.
+  '';
```
Comment (GaetanLepage) [inline plugins/by-name/wrapping/default.nix:20]:
```suggestion
  settingsOptions = {
    notify_on_switch = defaultNullOpts.mkBool true ''
       By default, wrapping.nvim will output a message to the
       command line when the hard or soft mode is set.
    '';
  };
```
Even if there is only a single option for now, this leaves us the room to add more later (purely cosmetic).

File: tests/test-sources/plugins/by-name/wrapping/default.nix [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{ pkgs, ... }:
+{
+  empty = {
+    plugins.wrapping.enable = true;
+  };
+
+  defaults.plugins.which-key = {
+    enable = true;
+    notify_on_switch = true;
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/by-name/wrapping/default.nix:9]:
```suggestion
    settings = {
      notify_on_switch = true;
    };
```
This option will be created within the `settings` option.

File: tests/test-sources/plugins/by-name/wrapping/default.nix [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  };
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/by-name/wrapping/default.nix:10]:
```suggestion
  defaults = {
    plugins.which-key = {
      enable = true;
      notify_on_switch = true;
    };
  };
```
Purely cosmetic, for consistency with the rest of the code base.

**Review by GaetanLepage — DISMISSED** [review]
Thanks a lot for giving it a shot !
This looks pretty good already.

I made a few comments.

**Review by ZainKergaye — COMMENTED** [review]
Thanks! As I go through learning more nix tricks I plan to add to this one plugin and hopefully go from there!

File: lib/maintainers.nix [hunk 6]
```diff
@@ -100,4 +100,10 @@
     githubId = 6452260;
     name = "Andrew Plaza";
   };
+  ZainKergaye = {
+    email = "zain@zkergaye.me";
+    github = "zainkergaye";
+    githubId = 62440012;
+    name = "Zain Kergaye";
+  };
```
Comment (MattSturgeon) [inline lib/maintainers.nix:108]: This should be done in a separate commit, titled `maintainers: add ZainKergaye`

File: tests/test-sources/plugins/by-name/wrapping/default.nix (hunk 4 above)
Comment (MattSturgeon) [inline tests/test-sources/plugins/by-name/wrapping/default.nix:9]:
```suggestion
    settings.notify_on_switch = true;
```

or

```suggestion

    settings = {
      notify_on_switch = true;
    };
```

**Review by MattSturgeon — COMMENTED** [review]
Looks good

**Comment by ZainKergaye** [comment]
I wasn't able to run `nix flake check --all-systems`, was it anything on my part? I seem to get the same error from some of the buildbot/nix-build tests. 


'''
       nixvim|wrapping-nvim ⇒ nix flake check --all-systems                      
       warning: ignoring untrusted substituter "https://nix-community.cachix.org", you are not a trusted user.
       Run "man nix.conf" for more information on the "substituters" configuration option.
       error:
       … while checking flake output "checks"

         at «none»:0: (source not available)

       … while checking the derivation "checks.x86_64-linux.docs"

         at «none»:0: (source not available)

       (stack trace truncated; use "--show-trace" to show the full trace)

       error: value is a set while a string was expected

'''

File: plugins/by-name/wrapping/default.nix [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  lib,
+  ...
+}:
+let
+  inherit (lib.nixvim) defaultNullOpts;
+in
+lib.nixvim.neovim-plugin.mkNeovimPlugin {
+  name = "wrapping";
+  originalName = "wrapping.nvim";
+  package = pkgs.vimPlugins.wrapping-nvim;
```
Comment (MattSturgeon) [inline plugins/by-name/wrapping/default.nix:12]:
```suggestion
  package = "wrapping-nvim";
```

**Review by MattSturgeon — COMMENTED** [review]
> I wasn't able to run `nix flake check --all-systems`, was it anything on my part? I seem to get the same error from some of the buildbot/nix-build tests.

I wouldn't recommend running that locally. It's a very heavy set of tests, 99% unrelated to this PR.

Instead you can use the `tests` command made available in the devshell, e.g. interactively with `tests -i` and filter for the relevant plugin's test.

In this case the issue seems to be:

File: tests/test-sources/plugins/by-name/wrapping/default.nix (hunk 4 above)
Comment (MattSturgeon) [reply tests/test-sources/plugins/by-name/wrapping/default.nix]: This is still not resolved:
```
error: The option `plugins.which-key.notify_on_switch' does not exist. Definition values:
- In `/nix/store/XXX-test-sources/plugins/by-name/wrapping/default.nix': true
```

**Comment by MattSturgeon** [comment]
> I wish the contributing.md would be a little more clear for me.

We'd love to hear any feedback you have on our docs! Perhaps you could open an issue or a discussion thread if you get chance?

> Running 'tests -i' and filtering out by plugins-by-name does not give me wrapping.

It's possible the devshell is outdated, maybe re-running `nix develop` will help? :thinking:

File: tests/test-sources/plugins/by-name/wrapping/default.nix
Comment (ZainKergaye) [inline tests/test-sources/plugins/by-name/wrapping/default.nix:1]: Probably should've looked at what I copy-pasted too haha

**Comment by ZainKergaye** [comment]
> > I wish the contributing.md would be a little more clear for me.
> 
> We'd love to hear any feedback you have on our docs! Perhaps you could open an issue or a discussion thread if you get chance?
> 
> > Running 'tests -i' and filtering out by plugins-by-name does not give me wrapping.
> 
> It's possible the devshell is outdated, maybe re-running `nix develop` will help? 🤔

I'm happy to do that! This is my real first attempt at contributing and I would love for others to get into this aswell!

Running 'nix-develop' worked! Thanks!

File: lib/maintainers.nix (hunk 6 above)
Comment (MattSturgeon) [reply lib/maintainers.nix:108]: This isn't resolved, but the commit history needs squashing anyway.

Once the PR is done, could you squash so that we only have two commits:
- `maintainers: add ZainKergaye`
- `plugins/wrapping: init`

File: tests/test-sources/plugins/by-name/wrapping/default.nix
Comment (MattSturgeon) [reply tests/test-sources/plugins/by-name/wrapping/default.nix:1]: As per https://github.com/nix-community/nixvim/pull/2325#discussion_r1778103109, you also need to nest the settings option under `settings`.

File: tests/test-sources/plugins/by-name/wrapping/default.nix (hunk 4 above)
Comment (ZainKergaye) [reply tests/test-sources/plugins/by-name/wrapping/default.nix]: Thanks, just did that and had fun running the specified tests :)

File: lib/maintainers.nix (hunk 6 above)
Comment (ZainKergaye) [reply lib/maintainers.nix:108]: How do I go about changing the commits?

Comment (MattSturgeon) [reply lib/maintainers.nix:108]: There's a few ways you can do it, e.g. with an "interactive" rebase. But in this case I would do:
```
# reset to the base commit without modifying the working tree
git reset --mixed cb2b76c1a9ec067ed0c449080f4973fecf8532ef

git add lib/maintainers.nix
git commit -m "maintainers: add ZainKergaye"

git add plugins
git add tests
git commit -m "plugins/wrapping: init"

git push --force-with-lease
```

**Comment by MattSturgeon** [comment]
Looks like a formatting error slipped through: https://buildbot.nix-community.org/#/builders/448/builds/1513

You can resolve this with:

```
nix fmt
git add plugins tests
git commit --amend
git push --force-with-lease
```

**Comment by ZainKergaye** [comment]
Thank you for the guidence, what is the general process from here now?

**Review by GaetanLepage — APPROVED** [review]
LGTM, well done @ZainKergaye !

**Review by MattSturgeon — APPROVED** [review]
LGTM, thanks for an excellent first PR!

Our `main` branch has had a few commits added since the base of this PR, so if I ask @mergifyio to merge it'll first rebase your commits (they'll show as authored by you and committed by @nix-infra-bot).

Up to you, but if you want to avoid that clunkyness, you can do:
```
git fetch upstream
git rebase upstream/main
git push --force-with-lease
```
(assuming you have our repo setup as a remote named `upstream`)

**Comment by MattSturgeon** [comment]
@mergifyio queue

**Comment by khaneliman** [comment]
> > Running 'tests -i' and filtering out by plugins-by-name does not give me wrapping.
> 
> It's possible the devshell is outdated, maybe re-running `nix develop` will help? 🤔

Whenever a new module is added we need to regenerate the tests available so either nix develop or direnv reload.
